### PR TITLE
TSan CI

### PR DESCRIPTION
**Pull request by bjacob**

### CODE REVIEW

File: build_tools/buildkite/pipelines/trusted/postsubmit.yml [hunk 1]
```diff
@@ -27,3 +27,12 @@ steps:
       ./build_tools/buildkite/scripts/wait_for_pipeline_success.py \
         --annotate \
         build-runtime-cmake
+
+  - label: "Executing build-and-test-tsan"
+    commands: |
+      export BUILDKITE_ANNOTATION_CONTEXT="$${BUILDKITE_STEP_ID}"
+      export BUILDKITE_ACCESS_TOKEN="$(gcloud secrets versions access latest \
+          --secret=iree-buildkite-presubmit-pipelines)"
```
Comment (GMNGeoffrey) [inline build_tools/buildkite/pipelines/trusted/postsubmit.yml:35]: As of https://github.com/google/iree/pull/9367, this bit is no longer necessary

```suggestion
```

File: build_tools/buildkite/pipelines/trusted/presubmit.yml [hunk 2]
```diff
@@ -42,3 +42,18 @@ steps:
...
+    plugins:
+      - https://github.com/GMNGeoffrey/smooth-checkout-buildkite-plugin#24e54e7729:
+          repos:
+            - config:
+                - url: ${BUILDKITE_REPO}
+                  ref: ${CONFIG_FETCH_REF}
+    commands: |
+      export BUILDKITE_ANNOTATION_CONTEXT="$${BUILDKITE_STEP_ID}"
+      export BUILDKITE_ACCESS_TOKEN="$(gcloud secrets versions access latest \
+          --secret=iree-buildkite-presubmit-pipelines)"
```
Comment (GMNGeoffrey) [inline build_tools/buildkite/pipelines/trusted/presubmit.yml:56]:
```suggestion
```

File: build_tools/cmake/build_and_test_tsan.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Build the IREE project with CMake. Designed for CI, but can be run manually.
+# This is equivalent to the build_tools/cmake/rebuild.sh script except it
+# first deletes the build directory to deal with any faulty caching.
+
+set -x
+set -e
```
Comment (GMNGeoffrey) [inline build_tools/cmake/build_and_test_tsan.sh:14]:
```suggestion
set -euo pipefail
```

File: build_tools/cmake/build_and_test_tsan.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# IREE_BUILD_SAMPLES=OFF because samples assume embedded linker.
+./build_tools/cmake/clean_build.sh \
+  -DIREE_BUILD_TESTS=ON \
+  -DIREE_BUILD_SAMPLES=OFF \
+  -DIREE_BUILD_COMPILER=ON \
+  -DIREE_ENABLE_TSAN=ON \
+  -DIREE_BYTECODE_MODULE_ENABLE_TSAN=ON \
+  -DIREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER=ON
+
+ctest -j $(nproc) --timeout 900 --output-on-failure
```
Comment (GMNGeoffrey) [inline build_tools/cmake/build_and_test_tsan.sh:27]: Use build_tools/cmake/ctest_all.sh which sets `CTEST_PARALLEL_LEVEL`, respects filtering, etc.

File: build_tools/cmake/build_and_test_tsan.sh [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+# This is equivalent to the build_tools/cmake/rebuild.sh script except it
+# first deletes the build directory to deal with any faulty caching.
+
+set -x
+set -e
+
+# IREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER=ON because TSan doesn't work
+# with embedded linker.
+# IREE_BUILD_SAMPLES=OFF because samples assume embedded linker.
+./build_tools/cmake/clean_build.sh \
```
Comment (GMNGeoffrey) [inline build_tools/cmake/build_and_test_tsan.sh:19]: I think you want to just invoke CMake directly. That script is doing some weird implicit stuff...

File: build_tools/cmake/build_and_test_tsan.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Build the IREE project with CMake. Designed for CI, but can be run manually.
+# This is equivalent to the build_tools/cmake/rebuild.sh script except it
+# first deletes the build directory to deal with any faulty caching.
```
Comment (GMNGeoffrey) [inline build_tools/cmake/build_and_test_tsan.sh:11]: This comment looks like copypasta? Should reference tsan instead

```suggestion
# Build the IREE project using CMake with Thread Sanitizer enabled and
# then runs all tests.
```

**Comment by bjacob** [comment]
Ah, the conflict that this runs into is informative - I was wondering how to do the test vs build part and this seems to answer it. Updating the PR soon.

**Review by GMNGeoffrey — COMMENTED** [review]
Thanks for trying this out! I've just set it so the presubmit pipeline will run on your PR, so that should happen on the next push and you can see how it works

File: build_tools/cmake/build_and_test_tsan.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# with embedded linker.
+# IREE_BUILD_SAMPLES=OFF because samples assume embedded linker.
+./build_tools/cmake/clean_build.sh \
+  -DIREE_BUILD_TESTS=ON \
+  -DIREE_BUILD_SAMPLES=OFF \
+  -DIREE_BUILD_COMPILER=ON \
+  -DIREE_ENABLE_TSAN=ON \
+  -DIREE_BYTECODE_MODULE_ENABLE_TSAN=ON \
+  -DIREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER=ON
+
```
[the remark this answers is not in the document]
Comment (bjacob) [reply build_tools/cmake/build_and_test_tsan.sh]: Actually rebasing this I learned how to make it a dependent pipeline and I suppose that's better.

**Comment by bjacob** [comment]
Why is https://buildkite.com/iree/presubmit/builds/209 doing `build-runtime-cmake` but not `build-tsan` ?

**Comment by bjacob** [comment]
Got help from @GMNGeoffrey , i had to create that PR from a branch on the main repo for buildkite to pick up my config.  Closing this PR, the new PR is https://github.com/google/iree/pull/9371

**Pull request closed without merging**